docs/04-reference/winglang-spec.md

```diff
@@ -869,30 +869,223 @@ let f = (arg1: num, var arg2: num) => {
 
 ### 1.7 Optionality
 
-Symbol `?` can mark a type as optional.  
-Optionality means the value behind type can be either present or nil.
+Nullity (`null`, `undefined`, `nil`) is a primary source of bugs in software.
+Being able to guarantee that a value will never be null makes it easier to write safe
+code without constantly having to take nullity into account.
 
-Rules of optionality apply to the entire new container type of `type?` and not
-the value behind it (`type`).  
+We are going to try and design Wing without allowing users to explicitly assign or use the `nil`
+value. Instead, the language offers syntax to support the various use cases that arise when dealing
+with *optionality*.
 
-Using the `??` operator allows one to safely access the value behind an optional
-variable by always providing a default, in case the variable is nil. This forces
-a value to be present for the l-value (left hand side of the assignment operator
-and guarantees what's returned is type-stripped from its `?` keyword).
+Here's a quick summary of how optionality works in Wing:
 
-> ```TS
-> let x? = 44;
-> let y = x ?? 55;
-> ```
+* `x: T?` marks `x` as "optional of T". This means that `x` can store a value of type `T` but it can also be in a state where there is no value stored there.
+* To test if `x` has a value or not, use `x?`, which returns a `bool`. If x has a value, `x?` will be `true`, otherwise it will be `false`.
+* The `if let y = x { } else { }` is a special if statement that binds `y` inside the block only if `x` includes a value. Otherwise, the `else` block will be executed.
+* The `x?.y` notation can be used to access object fields only if there is a value. It returns an undefined optional if the value is not defined.
+* The `x ?? y` notation will return the value in `x` (if there is one) or `y` otherwise (`y` must be of type `T`, not `T?`).
+* A default value can also be supplied using `= y` in the declaration of the field or argument, which implies that the type will always be defined.
 
-<details><summary>Equivalent TypeScript Code</summary>
+#### 1.7.1 Declaration
 
-> ```TS
-> const x: number? = 44;
-> const y: number = x ?? 55;
-> ```
+##### 1.7.1.1 Declaration in struct fields
 
-</details>
+```js
+struct Person {
+  name: str;
+  address: str?;
+}
+```
+
+In the above struct, the `address` field, the `foo` argument and the `my_opt` field are all marked
+as optionals using `?`.
+
+This means, for example, that a `Person` can be initialized without defining the `address`
+field:
+
+```js
+let my_person = Person { name: "david" };
+assert(my_person.address? == false);
+```
+
+It is also possible to use `=` in the field declaration to indicate the default value (similar to function arguments):
+
+```js
+struct FormatOpts {
+  radix: num = 10;
+}
+
+let opts = FormatOpts {};
+assert(opts.radix == 10);
+```
+
+##### 1.7.1.2 Declaration in variables
+
+Use `T?` to indicate that a variable is optional:
+
+```js
+let var y: num?;
```

Comment:
This is a violation of another section of the spec -- see the section on declarations. Every `let` statement must have `= <value>` clause.

(I'm totally happy with changing this or saying optionals are an exception, just trying to keep us honest 🙂)

Reply:
It makes sense to me that `let var` of an optional s won't require an initial value. There are many use cases for this. Let me check.

Reply:
I don't like this implicit nullness. I believe that very early on we'll have a case where I just want to say that there's nothing inside an optional and we'll need some syntax for that. I'm for not using the words null/nil/undefined just so it's clear. Good candidates are none/nothing/empty.

There's also this issue:
`let func = (x: num?) => { do something };`
does the argument `x` have a default "empty" value or is it a regular positional argument?

Reply:
> There's also this issue: `let func = (x: num?) => { do something };` does the argument `x` have a default "empty" value or is it a regular positional argument?

My current mental model of how this will work in the language/compiler is `x` will be type `num?` (that is, when you try using it inside the function, you need unwrap the value in some way) -- but a `num?` with a default value of nil/empty. Thus it's possible to call func() and it will still work.

Reply:
> Thus it's possible to call func() and it will still work.

But what about a `let func = (x: num?, y:num?) => { do something };`?
When I call `func(7)` which of the two should I assign the `7` to?
I think default values only make sense for named (non-positional) argument.

Reply:
I think this is similar to how default arguments work in TypeScript and Python (correct me if I'm wrong)

Reply:
Added `= nil` so now every let must have a value, even if it's optional.